controllers/workspace/pruner_controller.go

```diff
@@ -0,0 +1,422 @@
+		return fmt.Errorf("ConfigMap %s does not contain key %s", config.CronJobScript, PrunerScriptKey)
+	}
+
+	return nil
+}
+
+func (r *PrunerReconciler) ensureCronJob(ctx context.Context, namespace string, config *controllerv1alpha1.CleanupCronJobConfig) error {
+	suspend := true
+	if config != nil && config.Enable != nil {
+		suspend = *config.Enable == false
+	}
+	image := PrunerImage
+	if config != nil && config.Image != "" {
+		image = config.Image
+	}
+	retainTime := PrunerRetainTime
+	if config != nil && config.RetainTime != nil && *config.RetainTime != 0 {
+		retainTime = strconv.FormatInt(int64(*config.RetainTime), 10)
+	}
+	configMapName := PrunerConfigMap
+	if config != nil && config.CronJobScript != "" {
+		configMapName = config.CronJobScript
+	}
+
+	cronJob := &batchv1.CronJob{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: namespace, Name: PrunerCronJobName}, cronJob); err != nil {
+		if apierrors.IsNotFound(err) {
+			return r.Create(ctx, prunerCronJob(namespace, suspend, image, retainTime, configMapName))
+		}
+		return err
+	}
+
+	needUpdate := false
+	// suspend
+	if cronJob.Spec.Suspend != nil && *cronJob.Spec.Suspend != suspend {
+		cronJob.Spec.Suspend = &suspend
+		needUpdate = true
+	}
+	// image
+	if cronJob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Image != image {
+		cronJob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Image = image
+		needUpdate = true
+	}
+	// retainTime
+	containerEnvs := &cronJob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Env
+	envs := map[string]string{
+		"RETAIN_TIME": retainTime,
+	}
+	for name, val := range envs {
+		found := false
+
+		for i, env := range *containerEnvs {
+			if env.Name == name {
+				found = true
+			}
+			if env.Value != val {
+				(*containerEnvs)[i].Value = val
+				needUpdate = true
+				break
+			}
+		}
+
+		if !found {
+			*containerEnvs = append(*containerEnvs, corev1.EnvVar{Name: name, Value: val})
+			needUpdate = true
+		}
+	}
+	// configMap
+	if cronJob.Spec.JobTemplate.Spec.Template.Spec.Volumes[0].VolumeSource.ConfigMap.LocalObjectReference.Name != configMapName {
+		cronJob.Spec.JobTemplate.Spec.Template.Spec.Volumes[0].VolumeSource.ConfigMap.LocalObjectReference.Name = configMapName
+		needUpdate = true
+	}
+	if needUpdate {
+		if err := r.Update(ctx, cronJob); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func prunerDefaultConfigMap(namespace string) *corev1.ConfigMap {
+	labels := resourceLabels()
+	labels[constants.DevWorkspaceWatchConfigMapLabel] = "true"
+
+	return &corev1.ConfigMap{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      PrunerConfigMap,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Data: map[string]string{
+			PrunerScriptKey: `#!/usr/bin/env bash
+current_time=$(date +%s)
+echo "Current time: ${current_time}"
+echo "RETAIN_TIME: ${RETAIN_TIME}"
+for namespace in $(oc get namespaces -o go-template='{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}' | grep ".-devspaces")
```

Comment:
Change this to `oc get namespaces -l app.kubernetes.io/component=workspaces-namespace -o go-template='{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}'`

Reply:
done!